Refactor environment

dVoorhout opened the pull request:
The environment has been split up to make it less of a godclass. Created interfaces for every feature of an entity: Moving, displaying, colliding is being done with a jmonkey interface. The new entities have their own responsibility by using the mentioned interfaces. Moving is done with use of MoveBehaviours which can be attached to a movable entity. level piece generation and camera management also have their own classes and responsibility now.

marcelkuipers committed:
added displayable

marcelkuipers committed:
added ICollidable class

marcelkuipers committed:
extended Displayable

dVoorhout committed:
Added interfaces for movable objects, created movableBehaviour class

dVoorhout committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

Conflicts:
	ContextDesktop/src/main/java/com/github/migi_1/Context/model/Displayable.java

dVoorhout committed:
fixed naming

marcelkuipers committed:
singleton assetManager

marcelkuipers committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

dVoorhout committed:
removed ICollidable, now using jmonkeys collidable, implemented commander

dVoorhout committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
added LevelPiece class

marcelkuipers committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
fixed unnecessary method

dVoorhout committed:
added more javadoc, used projectAssetManager to load commander model

dVoorhout committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
integrated new level pieces in environment

marcelkuipers committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

dVoorhout committed:
added carrier class, added killable interface for carrier

marcelkuipers committed:
added camera classes

marcelkuipers committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

dVoorhout committed:
javadoc, added id field for carrier

marcelkuipers committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

dVoorhout committed:
moved files to new folders, started platform class

dVoorhout committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
added VRCam in environment

marcelkuipers committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
fixed import mistake

dVoorhout committed:
filled platform class

dVoorhout committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
added fly cam in environment

marcelkuipers committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
added levelGenerator class

marcelkuipers committed:
level pieces generator in environment

dVoorhout committed:
added movable behaviour

dVoorhout committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
level piece generation in environment

marcelkuipers committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

dVoorhout committed:
refactored spatials

marcelkuipers committed:
resolve merge conflict

dVoorhout committed:
general cleanup, checkstyle, javadoc

dVoorhout committed:
checkstyle for commander and IDisplay

dVoorhout committed:
checkstyle for IKillable

dVoorhout committed:
checkstyle for IMovable

dVoorhout committed:
spellingsmistakes, checkstyle for platform

dVoorhout committed:
Camera javadoc/checkstyle

dVoorhout committed:
checkstyle/cleanup Environment

dVoorhout committed:
more Environment cleanup, changed camera

dVoorhout committed:
checkstyle/javadoc for LevelGenerator

dVoorhout committed:
checkstyle/javadoc LevelPiece, also set vsync true to manage fps with speedups

dVoorhout committed:
Changed VRcam to normal camera, deleted VRcamera

NilsHullegien commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/Environment.java:
```diff
@@ -47,29 +42,27 @@
-    private static final int DIFFERENT_WORLDS = 5;
-
     private static final float STEERING_ANGLE = (float) (Math.sqrt(2.f) / 2.f);
 
```
PMD: Unused variable.

dVoorhout replied:
pmd is correct about this one, everything that has to do with steering and collisions will be refactored in a seperate branch to split up the amount of features in each branch.

NilsHullegien commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/Environment.java:
```diff
@@ -47,29 +42,27 @@
     private DirectionalLight sun;
     private DirectionalLight sun2;
 
     private float steering;
```
PMD: Unused variable

dVoorhout replied:
they are being added to the rootnode, which handles the rendering of these objects, pmd needs to shut it ;)

NilsHullegien replied:
Oh, sorry, I didn't realize that. In that case we will leave these PMD notifications for what it is.

NilsHullegien commented:
Besides my comments, the code itself looks good!

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/MovableBehaviour.java:
```diff
@@ -0,0 +1,20 @@
+ * @author Damian
+ *
+ */
+public class MovableBehaviour {
```
I would name this class MoveBehaviour, because now it sounds like it's a behaviour that can be moved, instead of a behaviour for moving.

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/MovableBehaviour.java:
```diff
@@ -0,0 +1,20 @@
+     */
+    public Vector3f getMoveVector() {
+        return new Vector3f(0, 0, 0);
+    }
```
I would make this class abstract and turn this method into an abstract method getMoveVector. Then have a class StaticMoveBehaviour (or a similar name) that returns Vector3f(0, 0, 0).

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/IDisplayable.java:
```diff
@@ -0,0 +1,39 @@
+import com.jme3.scene.Spatial;
+
+/**
+ * Interface for all items that are displayed.
```
all items that can\* be displayed.

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/IDisplayable.java:
```diff
@@ -0,0 +1,39 @@
+     * Moves the entity in the given direction.
+     * @param location of where to move the entity to.
+     */
+    void move(Vector3f location);
```
A spatial already has a move method. So forcing Displayables to implement this method is cruel. Instead what you can do is remove this method all together, or make it a default method and return getModel().move(location).

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/IDisplayable.java:
```diff
@@ -0,0 +1,39 @@
+     * Scale the object.
+     * @param f scale factor
+     */
+    void scale(float f);
```
The spatial alread has a scale method. ^^

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/IDisplayable.java:
```diff
@@ -0,0 +1,39 @@
+    void setModel(Spatial model);
+
+    /**
+     * Moves the entity in the given direction.
```
Moves the displayable*

It doesn't necessarily have to be an entity.

MrCrime commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/IDisplayable.java:
```diff
@@ -0,0 +1,39 @@
+    Spatial getModel();
+
+    /**
+     * Set the physical model of the object.
```
Sets*

In Javadoc, please refer to methods in third person instead of second :)

marcelkuipers committed:
renamed MovableBehaviour class

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/IMovable.java:
```diff
@@ -0,0 +1,25 @@
+public interface IMovable extends IDisplayable {
+
+    /**
+     * Method for gettin the movable behaviour of an entity.
```
getting*

marcelkuipers committed:
added StaticMoveBehaviour

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/IKillable.java:
```diff
@@ -0,0 +1,30 @@
+ * @author Damian
+ *
+ */
+public interface IKillable {
```
Shouldn't this interface also have an onKilled() method?
I would then have a collection of Killables in the Environment. Then in the update method of the Environment, in the update method check whether the health of the Killable is 0 or lower, then call the onKilled() method if it's true.

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/IMovable.java:
```diff
@@ -0,0 +1,25 @@
+     * Method for setting the movable behaviour of an entity.
+     * @param mbh the bahviour of the movement to set the entities behaviour to.
+     */
+    void setMovableBehaviour(MovableBehaviour mbh);
```
mbh isn't a good name.

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/IMovable.java:
```diff
@@ -0,0 +1,25 @@
+
+    /**
+     * Method for setting the movable behaviour of an entity.
+     * @param mbh the bahviour of the movement to set the entities behaviour to.
```
behaviour*

marcelkuipers committed:
solved merge conflicts

MrCrime commented:
We now have 3 classes that have a MoveBehaviour as their attribute, and there will be more once there are enemies and obstacles. So I recommend having an abstract Entity class that implements IMovable and has a MoveBehaviour as attribute.

If you then need a specific behaviour for a certain class, in the constructor of that class you can then call setBehaviour(myBehaviour). 
In the constructor of the Entity class you could also do something like moveBehaviour = new StaticMoveBehaviour(), so that every enemy doesn't move by default unless otherwise specified.

marcelkuipers committed:
renamed AccelerometerMoveBehaviour class

marcelkuipers committed:
fixed comment

marcelkuipers committed:
added default move and scale methods

marcelkuipers committed:
added onKilled method

marcelkuipers committed:
renamed variable

marcelkuipers committed:
fixed some checkstyle

marcelkuipers committed:
added final comments and fixed checkstyle

marcelkuipers committed:
added Entity class

marcelkuipers committed:
updated entity subclasses

marcelkuipers committed:
added comments

marcelkuipers committed:
fixed comment

marcelkuipers committed:
deleted annoying syso

marcelkuipers commented:
Thanks for the feedback, Luka! I agreed with you on all points and immediately updated the project.
The usage of default methods is really nice :)

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/Entity.java:
```diff
@@ -0,0 +1,31 @@
+ * @author Marcel
+ *
+ */
+public abstract class Entity {
```
Entity should be the one that implements IMovable, so that the other ones don't have to.

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/Entity.java:
```diff
@@ -0,0 +1,31 @@
+    /**
+     * Move behaviour of particular Entity.
+     */
+    protected MoveBehaviour moveBehaviour;
```
Because there is a getter and a setter for this moveBehaviour, you can just make it private.

MrCrime commented:
Also, by not having a constructor in Entity that initialises the moveBehaviour to something default like a StaticMoveBehaviour, you create the big danger of subclasses forgetting to set the moveBehaviour, causing a NullPointerException later on.

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/Carrier.java:
```diff
@@ -0,0 +1,117 @@
+ * @author Damian
+ *
+ */
+public class Carrier extends Entity implements IMovable, Collidable, IKillable {
```
Doesn't have to implement IMovable if Entity does it.

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/Commander.java:
```diff
@@ -0,0 +1,66 @@
+ * @author Damian
+ *
+ */
+public class Commander extends Entity implements IMovable, Collidable {
```
Doesn't have to implement IMovable if Entity does it.

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/model/entity/Platform.java:
```diff
@@ -0,0 +1,50 @@
+ * @author Damian
+ *
+ */
+public class Platform extends Entity implements IMovable {
```
Doesn't have to implement IMovable if Entity does it.

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/utility/ProjectAssetManager.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * Object instance.
+     */
+    private  static ProjectAssetManager instance;
```
should be final.

marcelkuipers committed:
entity now implements interface

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/utility/ProjectAssetManager.java:
```diff
@@ -0,0 +1,58 @@
+    public static ProjectAssetManager getInstance() {
+        if (instance == null) {
+            instance = new ProjectAssetManager();
+        }
```
Instead of initialising the instance here, you can do it on the same line as the attribute declaration:

private static final ProjectAssetManager intance = new ProjectAssetManager()

Doing this gets rid of the if statement.

LMiljak commented on ContextDesktop/src/main/java/com/github/migi_1/Context/utility/ProjectAssetManager.java:
```diff
@@ -0,0 +1,58 @@
+     */
+    public void setAssetManager(AssetManager assetManager) {
+        this.assetManager = assetManager;
+    }
```
Do we need more than one assetManager? (seriously I don't know)

And if no:
Unless setAssetManager has been called somewhere, getAssetManager will return null. Don't we want a default assetManager?

marcelkuipers committed:
updated Entity class

marcelkuipers committed:
added comment'

marcelkuipers committed:
updated ProjectAssetManager

MrCrime commented:
Also:

All entities are Displayable, and a method that will be the same for ALL entities is the setModel one. Just like with the behaviour remark I had, can you do the same with the model?

So having a model attribute in Entity. Then getModel and setModel just use the one in Entity. 
You can also then do is make an abstract getDefaultModel() in Entity, so the subclasses have to return their default model.

Why is this useful you ask? It eliminates having to implement setModel() in every Entity.

MrCrime commented:
You accidently removed the private constructor in ProjectAssetManager :)

marcelkuipers committed:
added private singleton constructor

MrCrime commented:
PS Marcel, I know you need this code for your Obstacles feature. What you can do is just make a branch from this one, so you don't have to wait until it's merged with master.

I mean, I still have posted my worst comment yet:
TESSSTS!
Yea I'm being a d*ck. But many of these classes are easily testable and since our DoD states that this can't be merged until it's fully tested, I can't allow you to merge it yet :p 

Since Remi and Nils have offered themselves up for testing (see backlog), I suggest we just wait for them to arrive here :)

MrCrime commented:
Anyway, so apart from the tests now, I think it's all good.

marcelkuipers committed:
moved setModel and getModel methods

marcelkuipers committed:
fixed checkstyle

marcelkuipers commented:
I have already made a branch that pulls code from this branch and even a branch that pulls code from that branch :P.

marcelkuipers commented:
Testing might take some time, so I am gonna close this PR for now.

marcelkuipers committed:
added Entity test class

Commit:
Camera class tested

marcelkuipers committed:
added getmovebehaviour test

marcelkuipers committed:
updated entity class

marcelkuipers committed:
changed test entity class

marcelkuipers committed:
added methods to abstract test class

Commit:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
added getMoveBehaviourTest

marcelkuipers committed:
added set moveBehaviour test

marcelkuipers committed:
added get and set model test

marcelkuipers committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

Commit:
Tested ProjectAssetManager class

Commit:
Added AssetManagerCreator singleton and levelpiece tested

Commit:
Updated pom for power mockito

Commit:
ContextMessages package 100%

marcelkuipers committed:
mocking projectassetmanager working

marcelkuipers committed:
added EntityTest again

marcelkuipers committed:
working tests in TestEntity

Commit:
Added cobertura since EclEmma isnt happy with PowerMockito

Commit:
Added cobertura in App and Messages packages

marcelkuipers committed:
added carrierTest

marcelkuipers committed:
carrierCollideWith test

marcelkuipers committed:
added takeDamage test for carrier

marcelkuipers committed:
onKilled test

marcelkuipers committed:
added tests for carrier

marcelkuipers committed:
cleaned up test classes

marcelkuipers committed:
added testCommander class

marcelkuipers committed:
AccelerometerTest

marcelkuipers committed:
staticMoveBehaviourTest

Commit:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
added comments

Commit:
LevelGenerator class tested

marcelkuipers committed:
fixed checkstyle

marcelkuipers committed:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

marcelkuipers committed:
updated testLevelPiece class and delete AssetManagerCreator

marcelkuipers committed:
fixed checkstyle

marcelkuipers committed:
modified levelGeneration class and test class

marcelkuipers committed:
updated levelGenerator test

Commit:
Environment Class tested

Commit:
Fixed javadoc and checkstyle/pmd

Commit:
Removed something that broke tests, readded it.

NilsHullegien commented:
Tests are added, so I'm reopening this PR.

NilsHullegien commented:
Closing the PR to fix the merge conflicts with master

Commit:
Changed some Javadoc in Entity class

Commit:
Merge with master

Commit:
Merge branch 'refactor_Environment' of https://github.com/MrCrime/ContextMigi-1 into refactor_Environment

Commit:
Refactored AcceleratorMoveBehaviour.

-Renamed it to ConstantSpeedMoveBehaviour, because this behaviour is
everything but accelerating.
-Instead of a default moveVector for the class (-0.2, 0, 0), the
moveVector should be passed in the constructor of the class.

Commit:
Changed Javadoc for Entity.

Commit:
Changed Javadoc for StaticMoveBehaviour

Commit:
Changed some Javadoc in IDisplayable

Commit:
Changed JavaDoc of IKillable.

Commit:
Made takeDamage a default method in IKillable.

Commit:
Removed the takeDamage method from Carrier

Commit:
Changed Javadoc in MoveBehaviour.

Commit:
Changed Javadoc in Carrier

Commit:
Removed the unused model field from TestEntity.

Commit:
Renamed TestAccelerator..... to TestConstantSpeed........

Commit:
Minor Javadoc changes in TestCarrier

Commit:
Minor changes to Javadoc in TestCommander class

Commit:
Some minor Javadoc changes in TestPlatform

MrCrime commented:
Nils fixed the conflict.
I did some minor changes that I felt obligated to do :) 
reopening it now.

MrCrime commented:
And I also immediately approve :+1: 
Even though I still disagree with the location of a Carrier being represented as an id, but ill fix that in another branch where Im working on the Accelerometer stuff.

MrCrime commented:
Isn't Github supposed to show the commits made since the last time the PR was opened? I can't see my/Nil's commits

Commit:
Attempt at making Github show commits.

MrCrime commented:
Okay that's very weird. I guess closing a branch is a bad idea since commits made since closing aren't being shown here. (You can ignore that last commit I made, I was hoping by pushing one new commit it would show the rest as well).

MrCrime commented:
(ignore that last commit please, I was being dumb)

NilsHullegien commented:
I approve of this code now as well, merging.
